debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+	if err != nil {
+		return "{}"
+	}
+	return string(b)
+}
+
+func (d *debugger) LaunchEval(ctx context.Context, props LaunchEvalProperties) (Session, error) {
+	store := inmem.New()
+	txn, err := store.NewTransaction(ctx, storage.TransactionParams{Write: true})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create store transaction: %v", err)
+	}
+
+	regoArgs := []func(*rego.Rego){
+		rego.Query(props.Query),
+		rego.Store(store),
+		rego.Transaction(txn),
+	}
+
+	if props.SkipOps == nil {
+		props.SkipOps = []topdown.Op{topdown.IndexOp, topdown.RedoOp, topdown.SaveOp, topdown.UnifyOp}
+	}
+
+	if props.EnablePrint {
+		regoArgs = append(regoArgs, rego.EnablePrintStatements(true),
+			rego.PrintHook(d.printHook))
+	}
+
+	if len(props.DataPaths) > 0 {
+		regoArgs = append(regoArgs, rego.Load(props.DataPaths, nil))
+	}
+
+	for _, bundlePath := range props.BundlePaths {
+		regoArgs = append(regoArgs, rego.LoadBundle(bundlePath))
+	}
+
+	if props.InputPath != "" && props.Input != nil {
+		return nil, fmt.Errorf("cannot specify both input and input path")
+	}
+
+	if props.Input != nil {
+		regoArgs = append(regoArgs, rego.Input(props.Input))
+	} else if props.InputPath != "" {
+		input, err := readInput(props.InputPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read input: %v", err)
+		}
+		regoArgs = append(regoArgs, rego.Input(input))
+	}
+
+	r := rego.New(regoArgs...)
+
+	pq, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare query for evaluation: %v", err)
+	}
+
+	if err := store.Commit(ctx, txn); err != nil {
+		return nil, fmt.Errorf("failed to commit store transaction: %v", err)
+	}
+
+	tracer := newDebugTracer()
+
+	vc := topdown.NewVirtualCache()
+
+	evalArgs := []rego.EvalOption{
+		rego.EvalRuleIndexing(true),
+		rego.EvalEarlyExit(true),
+		rego.EvalQueryTracer(tracer),
+		rego.EvalRuleIndexing(props.RuleIndexing),
+		rego.EvalVirtualCache(vc),
+	}
+
+	varManager := newVariableManager()
+	// Threads are 1-indexed.
+	t := newThread(1, "main", tracer, varManager, vc, store, d.logger)
+	s := newSession(ctx, d, varManager, props.LaunchProperties, []*thread{t})
+
+	go func() {
+		defer func() { _ = tracer.Close() }()
+		rs, evalErr := pq.Eval(s.ctx, evalArgs...)
+		if evalErr != nil {
+			var topdownErr *topdown.Error
+			if errors.As(evalErr, &topdownErr) && topdownErr.Code == topdown.CancelErr {
+				return
+			}
+			d.logger.Error("Evaluation failed: %v", evalErr)
+			return
+		}
+
+		tracer.resultSet = rs
+		s.result(t, rs)
+	}()
+
+	if err := s.start(); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func readInput(path string) (interface{}, error) {
+	path, err := fileurl.Clean(path)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	var input interface{}
+	if err := json.Unmarshal(data, &input); err != nil {
+		return nil, err
+	}
+
+	return input, nil
+}
+
+type session struct {
```

Comment:
Can this be accessed concurrently? I didn't see any locks so maybe not?

Reply:
This is the primary entrypoint for e.g. implementations of the DAP protocol, so we can't rule out concurrent access completely. It should be benign to add a global mutex. I'll look into fixing that.